### ISSUE TITLE
Add capability for non-inline grid labels to appear inside map

### ISSUE THREAD

**Issue by karlwx** [SciTools/cartopy#1642]
I want to be able to adjust the position of the grid tick labels to be inside the map boundary. Setting the gridliner xpadding works to move the labels away from the map, but setting the xpadding to a negative number results in hidden labels. I found that that non-inline labels that overlap the map boundary are hidden by default, so I experimented with setting the visibility of all labels to true. However, that resulted in the labels appearing at a 45-degree angle.

I'm essentially trying to do something [like this](https://stackoverflow.com/questions/47873389/matplotlib-move-tick-labels-inside-plot-area). Note: I am not looking to use the inline labels option, since (at least to my knowledge) there isn't a way to fix their positions to a particular x or y offset from the map boundaries. Changing the xpadding doesn't have any effect with the inline labels. Maybe this should be an option instead?

Example code:
```python
import matplotlib.pyplot as plt
import cartopy.crs as ccrs

plt.figure(figsize=(12,9))
ax = plt.axes(projection=ccrs.Miller())
ax.set_extent([-130,-65,20,55], crs=ccrs.PlateCarree())
ax.coastlines(resolution='50m')
gl = ax.gridlines(draw_labels=True, color='gray', linestyle='--')
gl.top_labels, gl.right_labels = False, False
gl.xlabel_style, gl.ylabel_style = {'weight': 'bold'}, {'weight': 'bold'}
#gl.xpadding = -10     # Ideally, this would move labels inside the map, but results in hidden labels
```
No xpadding:
![Screenshot from 2020-08-25 10-51-30](https://user-images.githubusercontent.com/45180714/91191431-c74d5780-e6c2-11ea-84a2-4f45720b50a3.png)

xpadding set to -10 and visibility hard-coded to True results in rotated labels:
![Screenshot from 2020-08-25 10-50-55](https://user-images.githubusercontent.com/45180714/91191434-c7e5ee00-e6c2-11ea-9842-3080a4c06aa5.png)

**Comment by greglucas** [comment]
A similar request was asked on gitter, so I'll add my comment here as well.
I think you need to set the keyword `rotate_labels` to False in your second example.

**Comment by karlwx** [comment]
I went back and did some experimenting again. For whatever reason, when I tried the same code again, now the labels are not rotated and I get the desired result:
![inmap_labels](https://user-images.githubusercontent.com/45180714/93896323-c8bc7080-fcbe-11ea-9d17-71facd29af36.png)
All I did was comment out the logic for determining if the labels overlapped the map in ```cartopy.mpl.gridliner._update_labels_visibility()```. 
I haven't experimented with every projection, but it would seem unnecessary to me to prevent non-gridline labels from being drawn inside the map boundary. I would think the xpadding and ypadding attributes would be enough to control this?

Interestingly, when I tried to pass ```rotate_labels=False``` to ```ax.gridlines()``` I get an error: 
```AttributeError: 'LineCollection' object has no property 'rotate_labels'```
It works as expected if I set it using ```gl.rotate_labels=False```.

**Comment by Hakkeh4x0r** [comment]
Looks great. I have set the rotate_label=False and padding -10, but labels are still disappearing. As Greg stated, i might need to set all text artist to set_visible(True). You have any idea to loop this?

**Comment by karlwx** [comment]
> Looks great. I have set the rotate_label=False and padding -10, but labels are still disappearing. As Greg stated, i might need to set all text artist to set_visible(True). You have any idea to loop this?

For me, it seems the label artists are only created after the figure has been displayed (at least where I'm playing with it, in a jupyter notebook). So I am unable to grab them and toggle visibility before the figure displays. 

If it's working differently for you, the following may work:
```python
for artist in gl.bottom_label_artists:
    artist.set_visible(True)
```
However, like I mentioned, this doesn't work for me as the label artists aren't created yet. My workaround involved modifying the cartopy source code.

**Comment by Hakkeh4x0r** [comment]
Ok, I tried the exact code that you provided above and it does not work at my end. The labels are still missing. What version of cartopy and matplotlib are you using? (actually all dependencies versions)

**Comment by karlwx** [comment]
Commenting out that block of code worked for me:
![Screenshot from 2020-09-23 10-01-49](https://user-images.githubusercontent.com/45180714/94023295-1353f080-fd84-11ea-8633-46f681dc0207.png)

cartopy 0.18.0
matplotlib 3.2.2

**Comment by Hakkeh4x0r** [comment]
After a good amount of tracing through the code, I noticed that the (as you mentioned) `_update_labels_visibility()` does perform the action setting visibility.

What I found was that if padding is set to negative value the Text Artists list is populated, but the difference is that the if-statement bellow checking intersection never triggers True. 

`
elif not outline_path.intersects_path(this_path):
    visible = True
`

which means that visibility is set to False for every Artist. 

`
artist.set_visible(visible)
`

I did some logging to see if I could find a clear difference in data populated, and;

- When padding is positive,

outline_path:
```
Path(array([
	[127.5, 110. ],
        [127.5, 880. ],
      	[897.5, 880. ],
      	[897.5, 110. ],
       	[127.5, 110. ]]), array([1, 2, 2, 2, 2], dtype=uint8))
```
this_path:
```
Path(array([
	[125.11104949,  66.        ],
       	[192.73604949,  66.        ],
      	[192.73604949,  80.        ],
      	[125.11104949,  80.        ],
       	[125.11104949,  66.        ]]), array([ 1,  2,  2,  2, 79], dtype=uint8))
```

_labels:
```
[('lon', True, Text(650918.2926992376, 4040085.029579938, "1°18'54''W")), ('lon', True, Text(650908.0582375554, 4040669.9627247, "1°18'54''W")), ('lon', True, Text(651187.0953597507, 4040085.029579938, "1°18'43''W")), ('lon', True, Text(651176.8426760917, 4040669.9627247, "1°18'43''W")), ('lon', True, Text(651455.8985829118, 4040085.029579938, "1°18'32''W")), ('lon', True, Text(651445.6276772781, 4040669.9627247, "1°18'32''W")), ('lat', True, Text(650894.4216919373, 4040284.104034253, "36°29'46''N")), ('lat', True, Text(651479.3548366983, 4040294.3567537377, "36°29'46''N")), ('lat', True, Text(650894.4216919373, 4040617.0199473985, "36°29'56''N")), ('lat', True, Text(651479.3548366983, 4040627.2737883506, "36°29'56''N"))]
```
- When padding is negative,

outline_path:
```
Path(array([
       [127.5, 110. ],
       [127.5, 880. ],
       [897.5, 880. ],
       [897.5, 110. ],
       [127.5, 110. ]]), array([1, 2, 2, 2, 2], dtype=uint8))
```

this_path:
```
Path(array([
       [125.11104949, 126.	],
       [192.73604949, 126.        ],
       [192.73604949, 140.        ],
       [125.11104949, 140.        ],
       [125.11104949, 126.        ]]), array([ 1,  2,  2,  2, 79], dtype=uint8))
```

_labels: 
```
[('lon', True, Text(650918.2926992376, 4040085.029579938, "1°18'54''W")), ('lon', True, Text(650908.0582375554, 4040669.9627247, "1°18'54''W")), ('lon', True, Text(651187.0953597507, 4040085.029579938, "1°18'43''W")), ('lon', True, Text(651176.8426760917, 4040669.9627247, "1°18'43''W")), ('lon', True, Text(651455.8985829118, 4040085.029579938, "1°18'32''W")), ('lon', True, Text(651445.6276772781, 4040669.9627247, "1°18'32''W")), ('lat', True, Text(650894.4216919373, 4040284.104034253, "36°29'46''N")), ('lat', True, Text(651479.3548366983, 4040294.3567537377, "36°29'46''N")), ('lat', True, Text(650894.4216919373, 4040617.0199473985, "36°29'56''N")), ('lat', True, Text(651479.3548366983, 4040627.2737883506, "36°29'56''N"))
```

I cant seem to figure out what to do next regarding this data. The data is changed at the second array position, but it doesn't tell me the meaning behind the number. Any clue?

What I did next while debugging was to manually set the parameter `visibility = True` for all iterations through Text Artists just to verify that the Text padding did work with negative number, and it did.

![test_image_23-13-07-15](https://user-images.githubusercontent.com/6514225/94058721-b1c26f00-fde1-11ea-8c09-f12f1598c494.png)

I tried climbing down the rabbit hole further,

```
this_patch = artist.get_bbox_patch()
this_path = this_patch.get_path().transformed(
    this_patch.get_transform())
```

but i'm a bit lost at this stage.

It seems that the Artist objects for all drawn Text labels are vanished when the gridliner object returns.

When I access the gl, fig or any object in debug mode there is no trace of the Text Artist or labels written.

@greglucas do you have any clue on how we could do this without tampering with the Cartopy code? I have as well some questions i would like to do privately, do have a email i could reach you on?

**Comment by stefraynaud** [comment]
Currently, non-inline labels are not drawn if they overlaps the map boundary polygon.
A solution is to only check against the map boundary polyline, so that labels that are inside are preserved as wall as those that are outside.

The label system is under heavy changes, and this may be added to the todolist.

**Comment by stefraynaud** [comment]
I made a little commit to the gridliner that allows the labels to be inside the map boundary.
I can find it on the **ksat** branch fo my fork : https://github.com/stefraynaud/cartopy/tree/ksat

Here is a sample based on the code above.
![ksat0](https://user-images.githubusercontent.com/1941408/94147044-0ebb3500-fe75-11ea-81ca-a8eb3da8b555.png)

**Comment by QuLogic** [comment]
Closed by #1537
